File: python/flink_agents/api/skills/agent_skill.py
#    Licensed to the Apache Software Foundation (ASF) under one
#   or more contributor license agreements.  See the NOTICE file
#   distributed with this work for additional information
#   regarding copyright ownership.  The ASF licenses this file
#   to you under the Apache License, Version 2.0 (the
#   "License"); you may not use this file except in compliance
#   with the License.  You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#  limitations under the License.
#

################################################################################
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
################################################################################
from typing import Dict, List

from pydantic import BaseModel, Field


class AgentSkill(BaseModel):
    """Represents an agent skill that can be loaded and used by agents.

    Attributes:
    ----------
    name : str
        Skill name (1-64 characters, lowercase letters, numbers, hyphens only).
        Must not start or end with a hyphen.
    description : str
        Skill description (1-1024 characters). Describes what the skill does
        and when to use it.
    license : Optional[str]
        License name or reference to a bundled license file.
    compatibility : Optional[str]
        Indicates environment requirements (intended product, system packages,
        network access, etc.) (max 500 characters).
    metadata : Optional[Dict[str, str]]
        Arbitrary key-value mapping for additional metadata.
    skill_content : str
        The skill implementation or instructions (the markdown body after
        frontmatter in SKILL.md).
    resources : Dict[str, str]
        Supporting resources referenced by the skill. Keys are relative paths
        from the skill root, values are file contents.
    """

    name: str = Field(..., min_length=1, max_length=64)
    description: str = Field(..., min_length=1, max_length=1024)
    content: str = Field(..., min_length=1)
    license: str | None = Field(default=None)
    compatibility: str | None = Field(default=None, max_length=500)
    metadata: Dict[str, str] | None = Field(default=None)
    resources: Dict[str, str] | None = None

    def get_resource(self, resource_path: str) -> str | None:
        """Get resource content by path.

        Args:
            resource_path: The relative path of the resource from skill root.

        Returns:
            The resource content, or None if not found.
        """
        return self.resources.get(resource_path)

    def get_resource_paths(self) -> List[str]:
        """Get all the resource relative paths of the skill."""
        if self.resources is None:
            return []
        return list(self.resources.keys())

File: runtime/src/main/java/org/apache/flink/agents/runtime/skill/AgentSkill.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.agents.runtime.skill;

import org.apache.flink.util.Preconditions;

import javax.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.function.Supplier;

/**
 * Runtime representation of one parsed {@code SKILL.md}.
 *
 * <p>Mirrors the Python {@code flink_agents.runtime.skill.agent_skill.AgentSkill}. Resources are
 * lazily loaded on first access.
 */
public final class AgentSkill {

    private final String name;
    private final String description;
    private final String content;
    @Nullable private final String license;
    @Nullable private final String compatibility;
    @Nullable private final Map<String, String> metadata;
    @Nullable private volatile Map<String, String> resources;
    @Nullable private Supplier<Map<String, String>> resourceLoader;
    private volatile boolean activated;

    public AgentSkill(
            String name,
            String description,
            String content,
            @Nullable String license,
            @Nullable String compatibility,
            @Nullable Map<String, String> metadata) {
        this(name, description, content, license, compatibility, metadata, null);
    }

    public AgentSkill(
            String name,
            String description,
            String content,
            @Nullable String license,
            @Nullable String compatibility,
            @Nullable Map<String, String> metadata,
            @Nullable Map<String, String> resources) {
        Preconditions.checkArgument(
                name != null && !name.isEmpty() && name.length() <= 64,
                "Skill name must be 1..64 characters: %s",
                name);
        Preconditions.checkArgument(
                description != null && !description.isEmpty() && description.length() <= 1024,
                "Skill description must be 1..1024 characters");
        Preconditions.checkArgument(
                content != null && !content.isEmpty(), "Skill content must not be empty");
        Preconditions.checkArgument(
                compatibility == null || compatibility.length() <= 500,
                "Skill compatibility must be at most 500 characters");
        this.name = name;
        this.description = description;
        this.content = content;
        this.license = license;
        this.compatibility = compatibility;
        this.metadata = metadata;
        this.resources = resources;
        this.activated = resources != null;
    }

    public String getName() {
        return name;
    }

    public String getDescription() {
        return description;
    }

    public String getContent() {
        return content;
    }

    @Nullable
    public String getLicense() {
        return license;
    }

    @Nullable
    public String getCompatibility() {
        return compatibility;
    }

    @Nullable
    public Map<String, String> getMetadata() {
        return metadata;
    }

    /** Set a lazy resource loader. Must be called before the first {@link #getResource(String)}. */
    public void setResourceLoader(Supplier<Map<String, String>> loader) {
        this.resourceLoader = loader;
    }

    /**
     * Return the content of the named resource (relative path from the skill root) or {@code null}
     * if no such resource is registered.
     */
    @Nullable
    public String getResource(String relativePath) {
        activate();
        return resources == null ? null : resources.get(relativePath);
    }

    /** Return all registered resource relative paths (sorted, may be empty). */
    public List<String> getResourcePaths() {
        activate();
        if (resources == null) {
            return List.of();
        }
        List<String> keys = new ArrayList<>(resources.keySet());
        keys.sort(String::compareTo);
        return keys;
    }

    private synchronized void activate() {
        if (!activated && resourceLoader != null) {
            resources = resourceLoader.get();
            activated = true;
        }
    }
}
